gst-ptp-helper needs to be able to open port < 1024
Using a full blown legacy suid would be risky, therefore
we modify it to use privileges.
Reported upstream as
https://gitlab.freedesktop.org/gstreamer/gstreamer/issues/511

diff --git a/libs/gst/helpers/gst-ptp-helper.c b/libs/gst/helpers/gst-ptp-helper.c
index 91a2570..a88e417 100644
--- a/libs/gst/helpers/gst-ptp-helper.c
+++ b/libs/gst/helpers/gst-ptp-helper.c
@@ -56,6 +56,10 @@
 #include <sys/capability.h>
 #endif
 
+#ifdef HAVE_PTP_HELPER_PRIVS
+#include <priv.h>
+#endif
+
 #include <glib.h>
 #include <gio/gio.h>
 
@@ -561,6 +565,10 @@ drop_privileges (void)
       g_error ("Failed to set process caps: %s", g_strerror (errno));
   }
 #endif
+#ifdef HAVE_PTP_HELPER_PRIVS
+  /* Remove PRIV_NET_PRIVADDR */
+  (void) priv_set(PRIV_OFF, PRIV_ALLSETS, PRIV_NET_PRIVADDR, NULL);
+#endif
 }
 
 static void
@@ -666,6 +674,11 @@ main (gint argc, gchar ** argv)
     g_error ("not running with superuser privileges");
 #endif
 
+#ifdef HAVE_PTP_HELPER_PRIVS
+  if (!priv_ineffect(PRIV_NET_PRIVADDR))
+    g_error ("not running with net_privaddr privilege");
+#endif
+
   opt_ctx = g_option_context_new ("- GStreamer PTP helper process");
   g_option_context_add_main_entries (opt_ctx, opt_entries, NULL);
   if (!g_option_context_parse (opt_ctx, &argc, &argv, &err))
diff --git a/libs/gst/helpers/meson.build b/libs/gst/helpers/meson.build
index b34423f..4dc53f6 100644
--- a/libs/gst/helpers/meson.build
+++ b/libs/gst/helpers/meson.build
@@ -105,6 +105,9 @@ if have_ptp
     endif
     cdata.set('HAVE_PTP_HELPER_CAPABILITIES', 1,
         description : 'Use capabilities for permissions in PTP helper')
+  elif with_ptp_helper_permissions == 'solaris-privileges'
+    cdata.set('HAVE_PTP_HELPER_PRIVS', 1,
+      description : 'Use Solaris privileges for permissions in PTP helper')
   else
     error('Unexpected ptp helper permissions value: ' + with_ptp_helper_permissions)
   endif
diff --git a/libs/gst/helpers/ptp_helper_post_install.sh b/libs/gst/helpers/ptp_helper_post_install.sh
index 4370acd..d5d58b3 100755
--- a/libs/gst/helpers/ptp_helper_post_install.sh
+++ b/libs/gst/helpers/ptp_helper_post_install.sh
@@ -20,7 +20,7 @@ case "$with_ptp_helper_permissions" in
     echo "Calling $setcap cap_net_bind_service,cap_net_admin+ep $ptp_helper"
     $setcap cap_net_bind_service,cap_net_admin+ep "$ptp_helper" || true
     ;;
-  none)
+  none|solaris-privileges)
     echo "No perms/caps to set for $ptp_helper"
     ;;
   *)
diff --git a/meson_options.txt b/meson_options.txt
index 72c3997..c5572ec 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -8,7 +8,7 @@ option('ptp-helper-setuid-user', type : 'string',
 option('ptp-helper-setuid-group', type : 'string',
        description : 'Group to switch to when installing gst-ptp-helper setuid root')
 option('ptp-helper-permissions', type : 'combo',
-       choices : ['none', 'setuid-root', 'capabilities', 'auto'], value : 'auto')
+       choices : ['none', 'setuid-root', 'capabilities', 'solaris-privileges', 'auto'], value : 'auto')
 option('option-parsing', type : 'boolean', value : true,
        description: 'Enable command line option parsing')
 option('poisoning', type : 'boolean', value : false, description : 'Enable poisoning of deallocated objects')
